### PR TITLE
stop retention user cleanup early again when DB connection attempt fails

### PR DESCRIPTION
**Pull request by FxKu**
In #2969 the syncSecrets function was updated to collect errors and continue processing the code. But for some errors like failing a connection attempt to the database we have to stop. We came across an edge case which produced retentionUsers during cluster creation that the operator then tried to query, but the database did not exist to this point.

Opening and closing the database connection has been moved inside the cleanupRotatedUsers function to allow for better error handling in SyncSecrets function.

I will add a test here for the edge case that we hit. It was a prepared database user conflicting with a manifest user of the same name except it uses - and not _. There should also be a better handling to not overwrite secrets when name conflicts exist.

### CODE REVIEW

File: pkg/cluster/sync.go [hunk 1]
```diff
@@ -1188,12 +1181,16 @@ func (c *Cluster) updateSecret(
 	} else {
 		// username might not match if password rotation has been disabled again
 		if secretUsername != string(secret.Data["username"]) {
-			*retentionUsers = append(*retentionUsers, secretUsername)
-			secret.Data["username"] = []byte(secretUsername)
-			secret.Data["password"] = []byte(util.RandomPassword(constants.PasswordLength))
-			secret.Data["nextRotation"] = []byte{}
-			updateSecret = true
-			updateSecretMsg = fmt.Sprintf("secret %s does not contain the role %s - updating username and resetting password", secretName, secretUsername)
+			if len(string(secret.Data["username"])) != len(secretUsername) {
```
Comment (mikkeloscar) [inline pkg/cluster/sync.go]: You already compare the user name strings in the if above, so no need to also compare the length? I assume same length but different content is equally wrong as different length.

Maybe you just need to add the else case which is not there in the current code.

File: pkg/cluster/sync.go [hunk 2]
```diff
@@ -1223,18 +1220,18 @@ func (c *Cluster) updateSecret(
 	if updateSecret {
 		c.logger.Infof("%s", updateSecretMsg)
 		if secret, err = c.KubeClient.Secrets(secret.Namespace).Update(context.TODO(), secret, metav1.UpdateOptions{}); err != nil {
-			return secret, fmt.Errorf("could not update secret %s: %v", secretName, err)
+			return secret, fmt.Errorf("could not update secret: %v", err)
```
[the remark this answers is not in the document]
Comment (FxKu) [reply pkg/cluster/sync.go]: the error handling of the UpdateSecret call already contains the secret name: `syncing secret %s failed: <err>`. Including the secret name again in the returned error felt redundant to me.

You're right about returning nil. Will update.

**Comment by mikkeloscar** [comment]
:+1:

**Comment by FxKu** [comment]
:+1:
